middleware/rate_limiter_test.go

```diff
@@ -0,0 +1,377 @@
+		id   string
+		code int
+	}{
+		{"127.0.0.1", http.StatusOK},
+		{"127.0.0.1", http.StatusOK},
+		{"127.0.0.1", http.StatusOK},
+		{"127.0.0.1", http.StatusForbidden},
+		{"", http.StatusBadRequest},
+		{"127.0.0.1", http.StatusForbidden},
+		{"127.0.0.1", http.StatusForbidden},
+	}
+
+	for _, tc := range testCases {
+		req := httptest.NewRequest(http.MethodGet, "/", nil)
+		req.Header.Add(echo.HeaderXRealIP, tc.id)
+
+		rec := httptest.NewRecorder()
+
+		c := e.NewContext(req, rec)
+
+		_ = mw(handler)(c)
+
+		assert.Equal(t, tc.code, rec.Code)
+	}
+}
+
+func TestRateLimiterWithConfig_defaultDenyHandler(t *testing.T) {
+	var inMemoryStore = NewRateLimiterMemoryStore(RateLimiterMemoryStoreConfig{rate: 1, burst: 3})
+
+	e := echo.New()
+
+	handler := func(c echo.Context) error {
+		return c.String(http.StatusOK, "test")
+	}
+
+	mw := RateLimiterWithConfig(RateLimiterConfig{
+		IdentifierExtractor: func(c echo.Context) (string, error) {
+			id := c.Request().Header.Get(echo.HeaderXRealIP)
+			if id == "" {
+				return "", errors.New("invalid identifier")
+			}
+			return id, nil
+		},
+		ErrorHandler: func(ctx echo.Context) error {
+			return ctx.JSON(http.StatusForbidden, nil)
+		},
+		Store: inMemoryStore,
+	})
+
+	testCases := []struct {
+		id   string
+		code int
+	}{
+		{"127.0.0.1", http.StatusOK},
+		{"127.0.0.1", http.StatusOK},
+		{"127.0.0.1", http.StatusOK},
+		{"127.0.0.1", http.StatusForbidden},
+		{"", http.StatusForbidden},
+		{"127.0.0.1", http.StatusForbidden},
+		{"127.0.0.1", http.StatusForbidden},
+	}
+
+	for _, tc := range testCases {
+		req := httptest.NewRequest(http.MethodGet, "/", nil)
+		req.Header.Add(echo.HeaderXRealIP, tc.id)
+
+		rec := httptest.NewRecorder()
+
+		c := e.NewContext(req, rec)
+
+		_ = mw(handler)(c)
+
+		assert.Equal(t, tc.code, rec.Code)
+	}
+}
+
+func TestRateLimiterWithConfig_defaultConfig(t *testing.T) {
+	{
+		var inMemoryStore = NewRateLimiterMemoryStore(RateLimiterMemoryStoreConfig{rate: 1, burst: 3})
+
+		e := echo.New()
+
+		handler := func(c echo.Context) error {
+			return c.String(http.StatusOK, "test")
+		}
+
+		mw := RateLimiterWithConfig(RateLimiterConfig{
+			Store: inMemoryStore,
+		})
+
+		testCases := []struct {
+			id   string
+			code int
+		}{
+			{"127.0.0.1", http.StatusOK},
+			{"127.0.0.1", http.StatusOK},
+			{"127.0.0.1", http.StatusOK},
+			{"127.0.0.1", http.StatusTooManyRequests},
+			{"127.0.0.1", http.StatusTooManyRequests},
+			{"127.0.0.1", http.StatusTooManyRequests},
+			{"127.0.0.1", http.StatusTooManyRequests},
+		}
+
+		for _, tc := range testCases {
+			req := httptest.NewRequest(http.MethodGet, "/", nil)
+			req.Header.Add(echo.HeaderXRealIP, tc.id)
+
+			rec := httptest.NewRecorder()
+
+			c := e.NewContext(req, rec)
+
+			_ = mw(handler)(c)
+
+			assert.Equal(t, tc.code, rec.Code)
+		}
+	}
+}
+
+func TestRateLimiterWithConfig_skipper(t *testing.T) {
+	e := echo.New()
+
+	var skipped bool
+	handler := func(c echo.Context) error {
+		skipped = true
```

Comment:
This is not the best way to check that the skipper worked as expected, because the handler is called if the skipper failed and the rest work as expected. A better way to assert it, is to put this boolean flag in the BeforeFunc and check that it remaind as false (the skipper worked and the BeforeFunc was not invoked)